### PR TITLE
Chore/split pending confirmations state

### PR DESCRIPTION
**Pull request by samsondav**
Fixes https://www.pivotaltracker.com/story/show/172003830

By resolving ambiguity between pending outgoing/incoming confirmations.

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -3919,7 +3919,7 @@
   resolved "https://registry.npmjs.org/@types/keygrip/-/keygrip-1.0.1.tgz#ff540462d2fb4d0a88441ceaf27d287b01c3d878"
   integrity sha1-/1QEYtL7TQqIRBzq8n0oewHD2Hg=
 
-"@types/lodash@^4.14.138", "@types/lodash@^4.14.139", "@types/lodash@^4.14.149":
+"@types/lodash@^4.14.138", "@types/lodash@^4.14.139":
```
Comment (j16r) [inline yarn.lock]: Are these changes related to this PR?

File: core/store/tx_manager.go [hunk 2]
```diff
@@ -555,12 +555,16 @@ func (txm *EthTxManager) CheckAttempt(txAttempt *models.TxAttempt, blockHeight u
 		return receipt, Unconfirmed, nil
 	}
 
+	// TODO: Could make this configurable i.e. by setting
+	// minOutgoingConfirmations param on the ethtx adapter and taking a max of
+	// that and the globally configured value as we do with
+	// MinRequiredIncomingConfirmations
```
Comment (j16r) [inline core/store/tx_manager.go]: Could this TODO be solved as part of this PR?

File: core/store/models/job_spec.go [hunk 3]
```diff
@@ -304,8 +304,9 @@ type Feeds = JSON
 // additional information that adapter would need to operate.
 type TaskSpec struct {
 	gorm.Model
-	JobSpecID     *ID           `json:"-"`
-	Type          TaskType      `json:"type" gorm:"index;not null"`
+	JobSpecID *ID      `json:"-"`
+	Type      TaskType `json:"type" gorm:"index;not null"`
+	// TODO: WTF is Confirmations here??
```
Comment (j16r) [inline core/store/models/job_spec.go]: Individual TaskSpecs can specify their required confirmations.

File: core/store/migrations/migration1585559482/migrate.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+// Split pending_confirmations into pending_outgoing_confirmations and pending_incoming_confirmations
+// For both task_runs and job_runs
+// This is inherently ambiguous (that's the entire reason for splitting them out) but we can make a best
+// guess by checking for presence of a transaction.
+func Migrate(tx *gorm.DB) error {
+	// HACK: Assume it's pending outgoing transactions if there exists a TX for the job run owning this task run
+	err := tx.Exec(`
+		UPDATE task_runs
```
Comment (j16r) [inline core/store/migrations/migration1585559482/migrate.go]: On some of our nodes we have many millions of TaskRuns, might want to double check this doesn't take too long as it will cause downtime.

File: core/adapters/eth_tx.go [hunk 5]
```diff
@@ -41,23 +41,23 @@ func (e *EthTx) TaskType() models.TaskType {
...
 // the blockchain.
-func (etx *EthTx) Perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+func (e *EthTx) Perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
 	if !store.TxManager.Connected() {
-		return pendingConfirmationsOrConnection(input)
+		return pendingOutgoingConfirmationsOrConnection(input)
 	}
 
-	if input.Status().PendingConfirmations() {
+	if input.Status().PendingOutgoingConfirmations() {
```
Comment (j16r) [inline core/adapters/eth_tx.go]: Personally I think this is the wrong way to go about solving this problem, it'd be better to look up transactions corresponding to this Task, rather than relying on a special obscure status for this individual task. It's clearer, simpler and less error prone.

e.g:

```go
tx, err := FindTxBySurrogateID(null.StringFrom(input.JobRunID().String()))
if errors.Cause(err) == errors.NotFound {
  // logic for creating initialTX
} else if err != nil {
 // error job
}

// process existing attempts logic
```

File: yarn.lock (hunk 1 above)
Comment (samsondav) [reply yarn.lock]: I do not understand why this is here now. I only ran `yarn`

File: core/store/tx_manager.go (hunk 2 above)
Comment (samsondav) [reply core/store/tx_manager.go]: Maybe, reading this now it doesnt even make sense so at the very least needs rewording

File: core/adapters/eth_tx.go (hunk 5 above)
Comment (spooktheducks) [reply core/adapters/eth_tx.go:49]: I can see your point, and it would make for significantly fewer changes to the code, but I'd offer a counterargument that Sam's approach has a few indirect benefits.  Curious what you think:

1. It decouples the system by ensuring that only one subsystem is responsible for determining which type of `PendingConfirmations` the run is in.  All other subsystems can just rely on the status handed to them.
    - If other subsystems end up needing to discern between the two, they don't need to reach out to the ORM for help.  Reaching out to the ORM involves I/O, and therefore more latency, which could (at least generally speaking) introduce races.
    - Separately, if a given subsystem only needs the ORM for this purpose and nothing else, we'll need to import the `orm` package.  I've seen some nasty import cycles pop up lately, and they can be pretty difficult to resolve cleanly, so I'm keen on avoiding them whenever possible.  Sure, you could make a one-method interface, but this is usually a sign of hastily-designed package architecture.

2. It might it easier to gather raw information from the DB about the state of a run for debugging if the status is explicitly recorded there.

Comment (j16r) [reply core/adapters/eth_tx.go:49]: I'll elaborate more on what the benefits are of my approach, and we can weigh them up against eachother:

  1. Checking for an existing TxAttempt record is the safest way to ensure no double spends occur, although I don't believe that's possible in the code, thar be dragons and I can't be 100% certain. 
  2. The RunManager has to make special exceptions for this code that complicate it, every other adapter accepts the nominal "InProgress" state for task+job, I also sense there are ways this status can be overwritten away from `pending_confirmations` that this code doesn't handle and again ... lead to a double spend.
  3. It's true this will require more DB calls, but there is another ticket which should make this a moot point: https://www.pivotaltracker.com/story/show/169746497
  4. It won't require a migration, which could be risky, and cause unnecessary downtime.

In my opinion double spends are kind of a trump card, so I want to design away any chance for them to happen.

Comment (j16r) [reply core/adapters/eth_tx.go:49]: Oh, and I forgot to mention: the suggestion to move most of this logic to the TxManager might solve some of the above ORM cycles/problems.

File: core/store/tx_manager.go (hunk 2 above)
Comment (samsondav) [reply core/store/tx_manager.go:561]: Yes, OK I remember now.

We can put a `minimumRequiredOutgoingConfirmations` parameter on either EthTX adapter or the Job Spec as a whole. Which is preferable? @j16r @se3000

Comment (samsondav) [reply core/store/tx_manager.go]: After a lot of faffing I've parked this work in another branch since it was more involved than expected. Card for it here: https://www.pivotaltracker.com/n/projects/2129823

File: core/adapters/eth_tx.go (hunk 5 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:49]: I think we can go with a combination of both approaches.

File: core/store/migrations/migration1585559482/migrate.go (hunk 4 above)
Comment (samsondav) [reply core/store/migrations/migration1585559482/migrate.go:14]: On our prod database, this takes under 100ms to run all statements.

However, I will address your other comment and perhaps a migration may not even be necessary.

File: core/adapters/eth_tx.go (hunk 5 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:49]: @j16r What about the `pending_connection` state. Do you think that ought to be eliminated as well?

**Comment by samsondav** [comment]
Current status on this:

Considering to replace with a different approach: https://www.pivotaltracker.com/story/show/172306435

Will keep this PR for now for reference.

**Comment by samsondav** [comment]
Re-opening as per discussion with @spooktheducks
